src/internal_rules.jl

```diff
@@ -447,6 +447,64 @@ function EnzymeRules.reverse(config, func::Const{typeof(\)}, ::Type{RT}, cache,
     return (nothing,nothing)
 end
 
+const EnzymeTriangulars = Union{
+    UpperTriangular,
+    LowerTriangular,
+    UnitUpperTriangular,
+    UnitLowerTriangular
+}
+
+function EnzymeRules.augmented_primal(
+    config,
+    func::Const{typeof(ldiv!)},
+    ::Type{RT},
+    Y::Annotation{YT},
+    A::Annotation{AT},
+    B::Annotation{BT}
+) where {RT, YT <: Array, AT <: EnzymeTriangulars, BT <: Array}
+    cache_Y = EnzymeRules.overwritten(config)[1] ? copy(Y.val) : Y.val
+    cache_A = EnzymeRules.overwritten(config)[2] ? copy(A.val) : A.val
+    cache_A = compute_lu_cache(cache_A, B.val)
+    cache_B = EnzymeRules.overwritten(config)[3] ? copy(B.val) : nothing
+    primal = EnzymeRules.needs_primal(config) ? Y.val : nothing
```

Comment:
Ah I see a problem here, you're never running the original function `ldiv!(A, B)`. Can you ensure this happens?

As well as make sure the tests confirm this?

Reply:
OMG. That is worrying! So does EnzymeTestUtils not check that the primal is the same as if the function is called?

Reply:
it appears not @sethaxen ?

Reply:
That's really strange. This line does check that the output of the primal and of the forward pass are the same:
https://github.com/EnzymeAD/Enzyme.jl/blob/cf0b4674a671c9c8f427e0a04a1c90a247b0adb2/lib/EnzymeTestUtils/src/test_reverse.jl#L129-L131

Reply:
I think it should be doing that:
https://github.com/simsurace/Enzyme.jl/blob/7ad702936e3907b5c582d5c35481e2fe4222a49c/lib/EnzymeTestUtils/src/test_reverse.jl#L129-L131

Reply:
> So the issue I think is that we need to check all of the in place arguments, not just the actual return. For example, here the issue is that the 1st arg is not mutated in place to have the return.

Aaaah, that's it. Yes we do test that an incorrect primal is detected in https://github.com/EnzymeAD/Enzyme.jl/blob/cf0b4674a671c9c8f427e0a04a1c90a247b0adb2/lib/EnzymeTestUtils/test/test_reverse.jl#L151-L162, but true, we need to do it for all arguments.

Reply:
Let me add a manual test of this so this PR doesn't get blocked by this refactor.